server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -654,7 +655,8 @@ private IndexMetadata(
         this.autoExpandReplicas = autoExpandReplicas;
         this.isSearchableSnapshot = isSearchableSnapshot;
         this.isPartialSearchableSnapshot = isPartialSearchableSnapshot;
-        this.indexCompatibilityVersion = SETTING_INDEX_VERSION_COMPATIBILITY.get(settings);
+        this.indexCompatibilityVersion = indexCompatibilityVersion;
+        assert indexCompatibilityVersion.equals(SETTING_INDEX_VERSION_COMPATIBILITY.get(settings));
```

Comment:
This snuck in recently making the constructor here needlessly heavy, fixed here so the copy-constructors are as fast as possible again.